frontend/model/contracts/group.js

```diff
@@ -107,6 +107,50 @@ function initFetchMonthlyPayments ({ meta, state, getters }) {
   return monthlyPayments
 }
 
+function insertMonthlyCycleEvent (state, event) {
+  // Loop through missing monthly cycle events that happen before the 'event' parameter's cycle
+  let lastEvent = state.distributionEvents[state.distributionEvents.length - 1]
+  while (Math.floor(lastEvent.data.cycle) !== Math.floor(event.data.cycle)) {
+    // Add the missing monthly cycle event
+    const monthlyCycleEvent = {
+      type: 'startCycleEvent',
+      data: {
+        cycle: lastEvent.data.cycle + 1,
+        latePayments: [] // List to be populated later, by the events-parser
+      }
+    }
+    state.distributionEvents.push(monthlyCycleEvent)
+    lastEvent = monthlyCycleEvent
+  }
+
+  state.distributionEvents.push(event)
+}
+
+function memberDeclaredIncome (state, username, haveNeed, createdDate) {
+  insertMonthlyCycleEvent(state, {
+    type: 'haveNeedEvent',
+    data: {
+      name: username,
+      haveNeed,
+      when: createdDate,
+      cycle: cycleAtDate(createdDate, state.distributionEvents[0].data.when)
+    }
+  })
+}
+
+function memberLeaves (state, username, dateLeft) {
+  state.profiles[username].status = PROFILE_STATUS.REMOVED
+  state.profiles[username].departedDate = dateLeft
+  insertMonthlyCycleEvent(state, {
+    type: 'userExitsGroupEvent',
```

Comment:
Do you need the `userExitsGroupEvent`? Or can you modify your algorithm to simplify it by adding a `haveNeedEvent` that sets the pledge to 0 for the user that left?

Reply:
I designed it this way because, eventually, `monthlyEventParser.js` needs to know who is present in order to handle mincome adjustments. Having a needer-pledger would cause this value of 0 to be adjusted with mincome changes.